druntime/src/core/internal/array/concatenation.d

```diff
@@ -8,71 +8,163 @@
 */
 module core.internal.array.concatenation;
 
-/// See $(REF _d_arraycatnTX, rt,lifetime)
-private extern (C) void[] _d_arraycatnTX(const TypeInfo ti, scope byte[][] arrs) pure nothrow;
+private template CatElems(string catSingleElem, string catArray)
+{
+    const char[] CatElems = q{
+        foreach (ref from; froms)
+            static if (is (typeof(from) : T))
+                } ~ catSingleElem ~ q{
+            else
+            } ~ "{\n" ~ catArray ~ "\n}\n";
+}
 
-/// Implementation of `_d_arraycatnTX` and `_d_arraycatnTXTrace`
-template _d_arraycatnTXImpl(Tarr : ResultArrT[], ResultArrT : T[], T)
+Tret _d_arraycatnTX(Tret, Tarr...)(scope auto ref Tarr froms) @trusted
```

Comment:
`froms` cannot be marked `scope` when it accepts array elements which will be copied into GC memory

Reply:
To elaborate, first of all add this to dmd/hdrgen.d to fix up `-vcg-ast:
```diff
--- a/compiler/src/dmd/hdrgen.d
+++ b/compiler/src/dmd/hdrgen.d
@@ -2336,6 +2336,19 @@ private void expressionPrettyPrint(Expression e, OutBuffer* buf, HdrGenState* hg
         expToBuffer(e.e2, cast(PREC)(precedence[e.op] + 1), buf, hgs);
     }
 
+    void visitCat(CatExp e)
+    {
+        if (e.lowering && global.params.vcg_ast)
+        {
+            expressionToBuffer(e.lowering, buf, hgs);
+            return;
+        }
+        else
+        {
+            visitBin(e);
+        }
+    }
+
     void visitComma(CommaExp e)
     {
         // CommaExp is generated by the compiler so it shouldn't
@@ -2639,6 +2652,7 @@ private void expressionPrettyPrint(Expression e, OutBuffer* buf, HdrGenState* hg
             return visit(e);
 
         case EXP.int64:         return visitInteger(e.isIntegerExp());
+        case EXP.concatenate:   return visitCat(e.isCatExp());
         case EXP.error:         return visitError(e.isErrorExp());
         case EXP.void_:         return visitVoidInit(e.isVoidInitExp());
         case EXP.float64:       return visitReal(e.isRealExp());
```
Then compile this with `-vcg-ast`:

```D
int[][] f() @safe 
{
    int[][] y;
    return y ~ [1, 2];
}
```

You'll see it gets lowered to:
```D
@safe int[][] f()
{
	int[][] y = null;
	return _d_arraycatnTX(y, ((int[2] __arrayliteral_on_stack3 = [1, 2];) , cast(int[])__arrayliteral_on_stack3));
}
```
Because you of the `scope`, the compiler feels free to put the array literal on stack, except `_d_arraycatnTX` escapes it by copying the stack pointer into the returned GC array resulting in a dangling pointer. The compiler doesn't catch this because `_d_arraycatnTX` is marked `@trusted`.

Reply:
I guess we have no alternative then to allocate the array literal with the gc (have it non-scope), and have the hook parameter as non-scope.

Reply:
Now I get it. I've removed the `scope` qualifier from the hook's argument.